XDMA/linux-kernel/xdma/xdma_netdev.c

```diff
@@ -0,0 +1,204 @@
+
+        desc->dst_addr_lo = cpu_to_le32(PCI_DMA_L(addr));
+        desc->dst_addr_hi = cpu_to_le32(PCI_DMA_H(addr));
+        desc->bytes = cpu_to_le32(len);
+}
+
+int xdma_netdev_open(struct net_device *ndev)
+{
+        struct xdma_private *priv = netdev_priv(ndev);
+        dma_addr_t dma_addr;
+        u32 w;
+
+        netif_carrier_on(ndev);
+        netif_start_queue(ndev);
+        /* Set the RX descriptor */
+        dma_addr = dma_map_single(
+                        &priv->xdev->pdev->dev,
+                        priv->res,
+                        sizeof(struct xdma_result),
+                        DMA_FROM_DEVICE);
+        priv->rx_desc->src_addr_lo = cpu_to_le32(PCI_DMA_L(dma_addr));
+        priv->rx_desc->src_addr_hi = cpu_to_le32(PCI_DMA_H(dma_addr));
+        rx_desc_set(priv->rx_desc, priv->dma_addr, XDMA_BUFFER_SIZE);
+        ioread32(&priv->rx_engine->regs->status_rc);
+
+        /* RX start */
+        w = cpu_to_le32(PCI_DMA_L(priv->rx_bus_addr));
+        iowrite32(w, &priv->rx_engine->sgdma_regs->first_desc_lo);
+
+        w = cpu_to_le32(PCI_DMA_H(priv->rx_bus_addr));
+        iowrite32(w, &priv->rx_engine->sgdma_regs->first_desc_hi);
+        iowrite32(DMA_ENGINE_START, &priv->rx_engine->regs->control);
+        return 0;
+}
+
+int xdma_netdev_close(struct net_device *ndev)
+{
+        netif_stop_queue(ndev);
+        netif_carrier_off(ndev);
+        return 0;
+}
+
+
+int check_desc1_status(struct xdma_private *priv, int index, dma_addr_t addr, u32 len, struct sk_buff *skb)
+{
+        unsigned long flag;
+        struct xdma_desc *desc = priv->desc[index];
+
+        spin_lock_irqsave(&priv->desc_lock[index], flag);
+        /* Need to check if the descriptor is empty */
+        if (desc->src_addr_lo == 0 && desc->src_addr_hi == 0) {
+                tx_desc_set(desc, addr, len);
+                priv->skb[index] = skb;
+                spin_unlock_irqrestore(&priv->desc_lock[index], flag);
+                return DESC_READY;
+        }
+        spin_unlock_irqrestore(&priv->desc_lock[index], flag);
+        return DESC_BUSY;
+}
+
+int check_desc2_status(struct xdma_private *priv, int index, dma_addr_t addr, u32 len, int desc1_status, struct sk_buff *skb)
+{
+        unsigned long flag;
+        struct xdma_desc *desc = priv->desc[index];
+
+        spin_lock_irqsave(&priv->desc_lock[index], flag);
+
+        /* Need to check if the descriptor is empty */
+        if (desc->src_addr_lo == 0 && desc->src_addr_hi == 0) {
+                if (desc1_status == DESC_READY) {
+                        spin_unlock_irqrestore(&priv->desc_lock[index], flag);
+                        return DESC_EMPTY;
+                }
+                tx_desc_set(desc, addr, len);
+                priv->skb[index] = skb;
+                spin_unlock_irqrestore(&priv->desc_lock[index], flag);
+                return DESC_READY;
+        }
+        spin_unlock_irqrestore(&priv->desc_lock[index], flag);
+        return DESC_BUSY;
+}
```

Comment:
What is different from `check_desc1_staus`?
It seems to be merged into single function and generalise to use more descriptors